File: day014_advanced_paradigms/related_single_number_xor/solution.py
import sys
from typing import List
from functools import reduce
import operator

class Solution:
    # Optimized: O(N) Time, O(1) Space - XOR Application
    def singleNumber_optimal(self, nums: List[int]) -> int:
        # XORing all elements together
        # A ^ A = 0, A ^ 0 = A. Pairs cancel out.
        return reduce(operator.xor, nums)

    # Brute Force: O(N) Time, O(N) Space - Hash Set (Fails O(1) space constraint)
    def singleNumber_brute(self, nums: List[int]) -> int:
        seen = set()
        for num in nums:
            if num in seen:
                seen.remove(num)
            else:
                seen.add(num)
        
        # The remaining element is the single number
        return seen.pop()

File: day014_advanced_paradigms/related_single_number_xor/solution.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <unordered_map>

using namespace std;

class Solution {
public:
    // Optimized: O(N) Time, O(1) Space - XOR Application
    int singleNumber_optimal(const vector<int>& nums) {
        int result = 0;
        // XOR all elements together
        for (int num : nums) {
            result ^= num;
        }
        return result;
    }

    // Brute Force: O(N) Time, O(N) Space - Hash Map/Set
    int singleNumber_brute(const vector<int>& nums) {
        unordered_map<int, int> counts;
        for (int num : nums) {
            counts[num]++;
        }
        for (auto const& [num, count] : counts) {
            if (count == 1) {
                return num;
            }
        }
        return -1; // Should not happen
    }
};